File: Signiture-Backend/kotlin/study/src/main/java/function/Exam.java
package function;

import java.util.List;
import java.util.function.Predicate;

public class Exam {

    public Exam() {
        var list = List.of(
                "1",
                "2",
                "홍길동",
                "함수",
                "메서드"
        );

        var stringPredicate = new Predicate<String>() {
            @Override
            public boolean test(String s) {
                return s.equals("?");
            }
        };

        List<String> result = list.stream()
                .filter(it -> it.equals("?"))
                .toList();
    }

    public static void main(String[] args) {
        new Exam();
    }
}

File: Signiture-Backend/kotlin/study/src/main/kotlin/function/Exam.kt
package pcy.study.function

import java.util.function.Predicate

fun main() {
    val list = listOf(
        "1",
        "2",
        "홍길동",
        "함수",
        "메서드"
    )

    val stringPredicate = object : Predicate<String> {
        override fun test(t: String): Boolean {
            return t == "?"
        }
    }

    list.filter { it == "?" }

    val add = { x: Int, y: Int -> x + y }
    println(add(3, 2))

    val higherOrderFunction = fun(x: Int, y: Int): Int {
        return x + y
    }
    println(higherOrderFunction(3, 2))

    lambda(5, 10, higherOrderFunction)
}

fun lambda(x: Int, y: Int, method: (Int, Int) -> Int) {
    println(method(x, y))
}
